### PR TITLE
Lookahead experimentation

### PR DESCRIPTION
**Pull request by BritikovKI**

### CODE REVIEW

File: src/smtsolvers/CoreSMTSolver.cc [hunk 1]
```diff
@@ -362,6 +363,13 @@ void CoreSMTSolver::attachClause(CRef cr)
     assert(c.size() > 1);
     watches[~c[0]].push(Watcher(cr, c[1]));
     watches[~c[1]].push(Watcher(cr, c[0]));
+    if(c.size() > 2 )
+        watches[~c[2]].push(Watcher(cr, c[0]));
+    else{
+        next_v[next_v.size() - 1].insert(var(~c[0]));
```
Comment (aehyvari) [inline src/smtsolvers/CoreSMTSolver.cc:369]: Maybe use `next_v.back().insert(var (~c[0]))` here and in similar places.

Comment (BritikovKI) [reply src/smtsolvers/CoreSMTSolver.cc]: Corrected

File: src/proof/PGBuild.cc [hunk 2]
```diff
@@ -307,16 +307,25 @@ void ProofGraph::buildProofGraph(const Proof & proof, int varCount) {
...
+#ifdef PROOFLOG
+        printf("Number of edges: %d\n", num_edges);
+#endif
         num_nodes = num_non_null;
 
         //reportf( "# Number of variables - nominal: %d - actual: %d\n",  num_vars_limit, proof_variables.size() );
         reportf("; Number of distinct variables in the proof: %d\n", (int) proof_variables.size());
+#ifdef PROOFLOG
+        printf("Number of distinct variables in the proof: %d\n", (int) proof_variables.size());
+#endif
```
Comment (aehyvari) [inline src/proof/PGBuild.cc:328]: These are all redundant. We shouldn't put them to the master.

File: src/smtsolvers/CoreSMTSolver.cc [hunk 3]
```diff
@@ -914,7 +914,7 @@ bool CoreSMTSolver::litRedundant(Lit p, uint32_t abstract_levels)
             if (cr == CRef_Fake)
             {
                 for (int j = top; j < analyze_toclear.size(); j++)
-                seen[var(analyze_toclear[j])] = 0;
+                    seen[var(analyze_toclear[j])] = 0;
```
Comment (aehyvari) [inline src/smtsolvers/CoreSMTSolver.cc:917]: This is a good catch!

File: src/smtsolvers/CoreSMTSolver.cc [hunk 4]
```diff
@@ -1603,7 +1603,7 @@ lbool CoreSMTSolver::search(int nof_conflicts, int nof_learnts)
                     // Model found:
                     return l_True;
             }
-            
+
```
Comment (aehyvari) [inline src/smtsolvers/CoreSMTSolver.cc:1606]: This is ok to change.

File: src/smtsolvers/CoreSMTSolver.h [hunk 5]
```diff
@@ -379,7 +379,7 @@ class CoreSMTSolver
     virtual void newDecisionLevel ();                                                  // Begins a new decision level.
     void     uncheckedEnqueue (Lit p, CRef from = CRef_Undef);                         // Enqueue a literal. Assumes value of literal is undefined.
     bool     enqueue          (Lit p, CRef from = CRef_Undef);                         // Test if fact 'p' contradicts current state, enqueue otherwise.
-    CRef     propagate        ();                                                      // Perform unit propagation. Returns possibly conflicting clause.
+    virtual CRef propagate    ();                                                      // Perform unit propagation. Returns possibly conflicting clause.
```
Comment (aehyvari) [inline src/smtsolvers/CoreSMTSolver.h:382]: Check indentation

**Review by aehyvari — COMMENTED** [review]
Looks good!

Some quick notes here: we shouldn't have experiments-related files in the master branch.  For example `parser.py` probably has no use outside of the paper.  Same for the class `CounterRewriter`.

I'm going to have a look at `LookaheadSMTSolver` next.

File: src/bin/Interpret.cc [hunk 6]
```diff
@@ -23,6 +23,8 @@ OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *********************************************************************/
 
+#include <cstdio>
+#include <CounterRewriter.h>
```
Comment (aehyvari) [inline src/bin/Interpret.cc:27]: I guess these are related to the experimentation.  We probably shouldn't put them to the master.

File: src/bin/Interpret.cc [hunk 7]
```diff
@@ -1307,7 +1309,11 @@ void Interpret::getInterpolants(const ASTNode& n)
         interpolationContext->getSingleInterpolant(itps, partitionings[0]);
     }
 
+    CounterRewriter interesting = CounterRewriter(*logic);
     for (int j = 0; j < itps.size(); j++) {
+        interesting.rewrite(itps[j]);
+        std::cout << "Amount of vars: " << interesting.config.varNumber <<
+            "\nAmount of boolean operators:" << interesting.config.boolNumber << "\n";
```
Comment (aehyvari) [inline src/bin/Interpret.cc:1316]: Experimentation stuff, shouldn't go to master.

File: src/rewriters/CounterRewriter.h [hunk 8]
```diff
@@ -0,0 +1,69 @@
+//
+// Created by Martin Blicha on 03.02.21.
+//
```
Comment (blishko) [inline src/rewriters/CounterRewriter.h:3]: Please update with proper license and correct author :)

File: src/rewriters/CounterRewriter.h [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+#ifndef OPENSMT_DIVMODREWRITER_H
```
Comment (blishko) [inline src/rewriters/CounterRewriter.h:5]: Use guard name corresponding to the file name.

File: src/rewriters/CounterRewriter.h [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+#include "Rewriter.h"
+#include "ArithLogic.h"
+
+#include "OsmtApiException.h"
+
+#include <unordered_map>
+#include <string>
+
+class CounterConfig : public DefaultRewriterConfig {
```
Comment (blishko) [inline src/rewriters/CounterRewriter.h:18]: It seems to me that this whole class should be a `Visitor`, not a `Rewriter`, no?

File: src/smtsolvers/CoreSMTSolver.cc (hunk 3 above)
Comment (blishko) [reply src/smtsolvers/CoreSMTSolver.cc:917]: Let's add parentheses!

Comment (BritikovKI) [reply src/smtsolvers/CoreSMTSolver.cc:917]: parentheses added

File: src/bin/Interpret.cc (hunk 7 above)
Comment (BritikovKI) [reply src/bin/Interpret.cc]: Removed

File: src/rewriters/CounterRewriter.h (hunk 10 above)
Comment (BritikovKI) [reply src/rewriters/CounterRewriter.h]: It's removed cause Antti said that it's not the thing to have in master

File: src/smtsolvers/LookaheadSMTSolver.h [hunk 11]
```diff
@@ -1,6 +1,9 @@
-//
-// Created by prova on 07.02.19.
-//
+/*
+ * Copyright (c) 2022, Antti Hyvarinen <antti.hyvarinen@gmail.com>
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.h:2]:
```suggestion
 * Copyright (c) 2019-2022, Antti Hyvarinen <antti.hyvarinen@gmail.com>
```

File: parser.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import sys
```
Comment (blishko) [inline parser.py:1]: Can we put this file into some reasonably name folder? Something like `utils`?
Also, maybe add some comment to the `main` method explaining what this actually does?

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 13]
```diff
@@ -1,16 +1,59 @@
-//
-// Created by prova on 07.02.19.
-//
+/*
+ * Copyright (c) 2022, Antti Hyvarinen <antti.hyvarinen@gmail.com>
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:2]:
```suggestion
 * Copyright (c) 2019-2022, Antti Hyvarinen <antti.hyvarinen@gmail.com>
```

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 14]
```diff
@@ -1,16 +1,59 @@
...
+ *
+ * SPDX-License-Identifier: MIT
+ */
 
 #include "LookaheadSMTSolver.h"
+#include "Proof.h"
+
+
+void LookaheadSMTSolver::attachClause(CRef cr) {
+    const Clause& c = ca[cr];
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:13]:
```suggestion
    Clause const & c = ca[cr];
```
Is the current style, I believe. What about that linter? :)

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 15]
```diff
@@ -239,120 +457,133 @@ std::pair<LookaheadSMTSolver::laresult,Lit> LookaheadSMTSolver::lookaheadLoop()
...
+                        return {laresult::la_tl_unsat, best}; // Problem is trivially unsat
+                    assert(checkTheory(true) == TPropRes::Decide);
 #ifndef NDEBUG
-                for (CRef cr : clauses) {
-                    Clause& c = ca[cr];
-                    unsigned k;
-                    for (k = 0; k < c.size(); k++) {
-                        if (value(c[k]) == l_True) {
-                            break;
+                    for (int j = 0; j < clauses.size(); j++) {
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:509]: Any reason why this has been turned from range-based for loop to classic for-loop? I don't see it.

**Review by blishko — COMMENTED** [review]
Left some minor comments. Otherwise it looks good to me, but I have not checked the changes in Lookahead much, I leave that to @aehyvari .

Also, I am strongly in favor of squashing these 135 commits, if nobody minds.

File: parser.py (hunk 12 above)
Comment (aehyvari) [reply parser.py:1]: I think the file is related to experiments.  We don't need to put it into master at all.

Comment (blishko) [reply parser.py:1]: Can we put it in `https://github.com/usi-verification-and-security/opensmt-debugging` ?
Seems like a useful thing to have available for the future.

Comment (aehyvari) [reply parser.py:1]: Sounds good to me!

File: src/tsolvers/THandler.cc [hunk 16]
```diff
@@ -442,6 +441,7 @@ Var     THandler::ptrefToVar         ( PTRef r ) { return tmap.getVar(r); }
 void    THandler::computeModel      () { getSolverHandler().computeModel(); } // Computes a model in the solver if necessary
 void    THandler::clearModel        () { /*getSolverHandler().clearModel();*/ }   // Clear the model if necessary
 
+bool    THandler::checkLitProp      (PtAsgn pta) { return getSolverHandler().wouldDeduce(pta); } // Push the assignment to all theory solvers
```
Comment (aehyvari) [inline src/tsolvers/THandler.cc:444]: The comment isn't up-to-date

File: src/tsolvers/THandler.h [hunk 17]
```diff
@@ -84,6 +84,7 @@ class THandler
     void    clearModel        ();// { /*getSolverHandler().clearModel();*/ }   // Clear the model if necessary
     bool    assertLits        (const vec<Lit> &);             // Give to the TSolvers the newly added literals on the trail
     bool    assertLit         (PtAsgn pta);// { return getSolverHandler().assertLit(pta); } // Push the assignment to all theory solvers
+    bool    checkLitProp      (PtAsgn pta);// { return getSolverHandler().assertLit(pta); } // Push the assignment to all theory solvers
```
Comment (aehyvari) [inline src/tsolvers/THandler.h:87]: The comment isn't up-to-date

File: src/tsolvers/TSolverHandler.cc [hunk 18]
```diff
@@ -24,6 +24,23 @@ void TSolverHandler::fillTheoryFunctions(ModelBuilder & modelBuilder) const {
     }
 }
 
+bool TSolverHandler::wouldDeduce(PtAsgn asgn){
+    bool res = true;
+    // Push backtrack points and the assignments to the theory solvers
+    // according to the schedule
+    for (int i = 0; i < solverSchedule.size(); i++) {
```
Comment (aehyvari) [inline src/tsolvers/TSolverHandler.cc:31]: Use range-based iteration here.

File: src/tsolvers/TSolverHandler.cc [hunk 19]
```diff
@@ -24,6 +24,23 @@ void TSolverHandler::fillTheoryFunctions(ModelBuilder & modelBuilder) const {
...
+        int idx = solverSchedule[i];
+        assert(tsolvers[idx] != nullptr);
+        tsolvers[idx]->pushBacktrackPoint();
```
Comment (aehyvari) [inline src/tsolvers/TSolverHandler.cc:34]: I don't believe that we need to push a backtrack point for this?

File: src/tsolvers/emptysolver/EmptySolver.cc [hunk 20]
```diff
@@ -135,6 +135,11 @@ void EmptySolver::computeModel( )
 {
 }
 
+
+bool EmptySolver::wouldDeduce( PtAsgn p ){
```
Comment (aehyvari) [inline src/tsolvers/emptysolver/EmptySolver.cc:139]: An unused variable `p`

File: src/tsolvers/lasolver/LASolver.cc [hunk 21]
```diff
@@ -413,7 +413,10 @@ void LASolver::getNewSplits(vec<PTRef>& splits) {
 bool LASolver::assertLit(PtAsgn asgn)
 {
     assert(asgn.sgn != l_Undef);
-
+    if(status != INIT){
```
Comment (aehyvari) [inline src/tsolvers/lasolver/LASolver.cc:416]: space between `if` and `(`

File: src/tsolvers/lasolver/LASolver.cc [hunk 22]
```diff
@@ -413,7 +413,10 @@ void LASolver::getNewSplits(vec<PTRef>& splits) {
...
+        bool assgn = wouldDeduce(asgn);
+        printf("Would assign: %d \n", assgn);
```
Comment (aehyvari) [inline src/tsolvers/lasolver/LASolver.cc:418]: Is this just for debugging?

File: src/tsolvers/lasolver/LASolver.cc [hunk 23]
```diff
@@ -638,10 +641,10 @@ void LASolver::checkDeduction() {
     printf("Checked deduction");
 }
 
-bool LASolver::wouldDeduce(PtAsgn asgn) const {
+bool LASolver::wouldDeduce(PtAsgn asgn) {
```
Comment (aehyvari) [inline src/tsolvers/lasolver/LASolver.cc:644]: Why can't this be `const`?

**Review by aehyvari — COMMENTED** [review]
This looks like the correct direction, or at least the one I had in mind now.  I believe it should work!  I left some comments while I was at it.

**Pull request closed without merging**